plotly/tools.py

```diff
@@ -2871,3 +2934,188 @@ def get_candle_decrease(self):
+                "rgb(40,35,35)",    # black
+                "rgb(133,20,75)",   # magenta
+                "rgb(255,65,54)",   # red
+                "rgb(255,255,255)",   # white
+                "rgb(255,220,0)"]   # yellow
+
+        for i in range(len(default_colors.keys())):
+            k = default_colors.keys()[i]
+            if i < len(colorscale):
+                default_colors[k] = colorscale[i]
+                
+        return default_colors
+   
+    def set_axis_layout(self, axis_key):
+        """ 
+            Sets and returns default axis object for dendrogram figure
+            :param (str) axis_key: "xaxis", "xaxis1", "yaxis", yaxis1", etc.
+        """
+
+        axis_defaults = {
+                'type': 'linear',
+                'ticks': 'outside',
+                'mirror': 'allticks',
+                'rangemode': 'tozero',
+                'showticklabels': True,                 
+                'zeroline': False,
+                'showgrid': False,              
+                'showline': True,
+            }
+        
+        if len(self.labels) != 0:
+            axis_key_labels = self.xaxis
+            if self.orientation in ['left', 'right']:
+                axis_key_labels = self.yaxis
+            if axis_key_labels not in self.layout:
+                self.layout[axis_key_labels] = {}
+            self.layout[axis_key_labels]['tickvals'] = [ea*self.sign[axis_key]
+                                                        for ea in self.zero_vals]
+            self.layout[axis_key_labels]['ticktext'] = self.labels
+            self.layout[axis_key_labels]['tickmode'] = 'array'
+            
+        self.layout[axis_key].update(axis_defaults)
+            
+        return self.layout[axis_key]
+
+    def set_figure_layout(self, width, height):
+        """ 
+            Sets and returns default layout object for dendrogram figure 
+        """
+        
+        self.layout.update({
+            'showlegend': False,
+            'autoscale': False,         
+            'hovermode': 'closest',
+            'width': width,
+            'width': height
+        })
+        
+        self.set_axis_layout(self.xaxis)
+        self.set_axis_layout(self.yaxis)
+
+        return self.layout
+
+    def get_dendrogram_traces(self, X, colorscale):
+        """ 
+            Calculates all the elements needed for plotting a dendrogram
+
+            :rtype (tuple): Contains all the traces in the following order
+            (a) trace_list: List of Plotly trace objects for the dendrogram
+            tree
+            (b) icoord: All X points of the dendogram tree as array of arrays
+            with length 4
+            (c) dcoord: All Y points of the dendogram tree as array of arrays
+            with length 4
+            (d) ordered_labels: leaf labels in the order they are going to
+            appear on the plot
+            (e) P['leaves']: left-to-right traversal of the leaves
+        """
```

Comment:
:cow2: again, I think you can save your space the indentation in these docstrings by just unindenting the whole thing.